umbraco/Umbraco-CMS
# Document and document type read API

### Prerequisites


### Description

This PR adds read operations for document and document types to the management API.

A few usages of the obsolete `ILocalizationService` have also been replaced with `ILanguageService` because I stumbled upon them while doing the PR 🤓 

### Limitations

Document scheduled publishing and document type list view configuration are conscioulsy omitted from this PR because their data contracts are yet to be defined.

### Breaking changes

We need to break `ContentTypeSort` to add document type keys for the management API output related to allowed child document types.

Eventually we will break it once more when we remove the integer ID, but this cannot be done until we get rid of the content type editor in the current backoffice.

### Special quirkyness

The core currently uses `Json.NET` for serialization, but the management API uses `System.Text.Json`. This causes some issues when reading document properties based on JSON, because the two serializers clash horribly when attempting to serialize the document output from the API.

Eventually we will replace `Json.NET` entirely for V13, but since we are still very much dependent on the current backoffice to create test data, we can't make that switch just yet as it would break the current backoffice. 

To maintain backwards compatability with the current backoffice while still being able to utilize `System.Text.Json` for serializing core data (i.e. document properties), this PR introduces the `ContextualJsonSerializer`. This serializer implementation attempts to be clever about the current request state, thus using `System.Text.Json` for serializing when the current request is within the management API and `Json.NET` otherwise. 

This is purely a temporary measure. It might not work flawlessly - there may be cases where the core ends up caching `System.Text.Json` data in memory and thus breaking later on when attempting to use the cached data for `Json.NET` serialization. We'll have to cross that bridge when we get there. Hopefully it won't be a huge issue moving forwards.

### Testing this PR

1. Verify that documents can be read from `https://localhost:44331/umbraco/management/api/v1.0/document/{key}`
2. Verify that document types can be read from `https://localhost:44331/umbraco/management/api/v1.0/document-type/{key}`
